### PR TITLE
[DI] Allow explicit taskLauncher for torchscript interpreter

### PR DESCRIPTION
**Pull request by garroud**
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/46362

By default, TorchScript execution is single threaded and uses the caller's thread pool. For the use case of distributed inference, we hope there is a way to customize the behavior where the  interpreter in torch script can be executed in other places. This diff allows an explicit taskLauncher for torchscript interpreter.

Test Plan: unit test is passed.

Differential Revision: D24114874

fbshipit-source-id: 1d7b003926c0d1f8facc53206efb960cff8897ac

Fixes #{issue number}

### CODE REVIEW

File: test/cpp/jit/test_graph_executor.cpp [hunk 1]
```diff
@@ -29,5 +32,40 @@ TEST(GraphExecutorTest, Basic_CUDA) {
   ASSERT_TRUE(almostEqual(stack[1].toTensor(), r1));
 }
 
+TEST(GraphExecutorTest, runAsync_executor) {
+  /*
+  TODO: there are some problem with C++ parsing script program involving
```
Comment (houseroad) [inline test/cpp/jit/test_graph_executor.cpp:37]: Do we have an issue on GIthub to track this issue? If so, we could add it in the comment.
If not, we could open one for tracking purpose.

File: test/cpp/jit/test_graph_executor.cpp [hunk 2]
```diff
@@ -29,5 +32,40 @@ TEST(GraphExecutorTest, Basic_CUDA) {
...
+        r2 = torch.jit.fork(torch.mm, torch.rand(100,100),torch.rand(100,100))
+        return r1.wait() + r2.wait()
+  demo = DemoModule()
+  torch.jit.save(torch.jit.script(demo), 'test_interpreter_async.pth')
+  */
+  std::string filePath(__FILE__);
+  auto testModelFile = filePath.substr(0, filePath.find_last_of("/\\") + 1);
+  testModelFile.append("test_interpreter_async.pt");
+  auto module = load(testModelFile);
+  module.to(at::kCPU);
```
Comment (houseroad) [inline test/cpp/jit/test_graph_executor.cpp:52]: wondering why we need to move the module to CPU, isn't the default device is cpu?

**Review by houseroad — APPROVED** [review]
Looks good.

**Review by suo — APPROVED** [review]
lgtm also. Is there a followup diff where we show how we expect to be using this customization point? Right now the highest-level place where we are addign hooks is `GraphExecutor`, which is not particularly user-facing (users generally interact with stuff in `jit/api/`, like `Module` or `Function`.

File: test/cpp/jit/test_graph_executor.cpp (hunk 2 above)
Comment (garroud) [reply test/cpp/jit/test_graph_executor.cpp:52]: it seems to be redundant, will remove.

**Pull request closed without merging**